GVFS/GVFS.Common/ProductUpgraderPlatformStrategy.cs

```diff
@@ -0,0 +1,38 @@
+using GVFS.Common.FileSystem;
+using GVFS.Common.Tracing;
+using System;
+using System.IO;
+
+namespace GVFS.Common
+{
+    public abstract class ProductUpgraderPlatformStrategy
+    {
+        public ProductUpgraderPlatformStrategy(PhysicalFileSystem fileSystem, ITracer tracer)
+        {
+            this.FileSystem = fileSystem;
+            this.Tracer = tracer;
+        }
+
+        protected PhysicalFileSystem FileSystem { get; }
+        protected ITracer Tracer { get; }
+
+        public abstract bool TryPrepareLogDirectory(out string error);
+
+        public abstract bool TryPrepareApplicationDirectory(out string error);
+
+        public abstract bool TryPrepareDownloadDirectory(out string error);
+
+        protected void TraceException(Exception exception, string method, string message)

```

Comment:
This is a useful utility method, that is not related to `ProductUpgraderPlatformStrategy`. Thoughts on moving this to `JsonTracer` as a new `RelatedException` method?

Reply:
Yeah - good point. We duplicate this utility method is several places, and we should probably centralize it. I will include it as a potential cleanup (and standardize this everywhere).